website/docs/async.md

```diff
@@ -0,0 +1,119 @@
+
+Given that a piece of code is worth more than a 1000 words, below you can find code snippets that explain these 
+semantics using known Scala technologies. In each of them `Output` is replaced with a respective async datatype
+to explain what internals of Besom are actually doing when resource constructors are called (oversimplified a 
+bit).
+
+<Tabs>
+  <TabItem value="Future" label="stdlib Future" default>
+
+```scala
+// internal function, here just to represent types
+def resolveResourceAsync(name: String, args: Args, promises: Promise[_]*): Future[Unit] = ???
+
+// resource definition
+case class Bucket(bucket: Future[String])
+object Bucket:
+  def apply(name: String, args: BucketArgs = BucketArgs()): Future[Bucket] = 
+    // create a promise for bucket property
+    val bucketNamePromise = Promise[String]() 
+    // kicks off async resolution of the resource properties
+    resolveResourceAsync(name, args, bucketNamePromise) 
+    // returns immediately
+    Future.successful(Bucket(bucketNamePromise.future)) 
+
+// this just returns a Future[Unit] that will be resolved immediately
+for 
+  a <- Bucket("first")
+  b <- Bucket("second")
+yield ()
+```
+
+  </TabItem>
+  <TabItem value="ce" label="Cats Effect IO">
+
+```scala
+// internal function, here just to represent types
+def resolveResourceAsync(name: String, args: Args, promises: Deferred[IO, _]*): IO[Unit] = ???
+
+// resource definition
+case class Bucket(bucket: IO[String])
+object Bucket:
+  def apply(name: String, args: BucketArgs = BucketArgs()): IO[Bucket] = 
+    for 
+      // create a deferred for bucket property
+      bucketNameDeferred <- Deferred[IO, String]() 
+      // kicks off async resolution of the resource properties
+      _ <- resolveResourceAsync(name, args, bucketNameDeferred).start 
+    yield Bucket(bucketNameDeferred.get) // returns immediately
+
+// this just returns a IO[Unit] that will be resolved immediately
+for 
+  a <- Bucket("first")
+  b <- Bucket("second")
+yield ()
+```
+
+  </TabItem>
+  <TabItem value="zio" label="ZIO">
+
+```scala
+// internal function, here just to represent types
+def resolveResourceAsync(name: String, args: Args, promises: Promise[_]*): Task[Unit] = ???
+
+// resource definition
+case class Bucket(bucket: Task[String])
+object Bucket:
+  def apply(name: String, args: BucketArgs = BucketArgs()): Task[Bucket] = 
+    for 
+      // create a promise for bucket property
+      bucketNamePromise <- Promise.make[Exception, String]() 
+      // kicks off async resolution of the resource properties
+      _ <- resolveResourceAsync(name, args, bucketNameDeferred).fork 
+    yield Bucket(bucketNameDeferred.await) // returns immediately
+
+// this just returns a Task[Unit] that will be resolved immediately
+for 
+  a <- Bucket("first")
+  b <- Bucket("second")
+yield ()
+```
+
+  </TabItem>
+</Tabs>
+
+There is a way to inform Pulumi engine that some of the resources have to be created, updated or deleted 
```

Comment:
It would be worth noting that "depends on" is needed only when Pulumi cannot infer the sequence from output dependency graph

Reply:
yes, this is exactly what I had in mind here:

draft of explanation of how to structure besom programs, I think we should expand upon this

but it feels like a big topic and I'm thinking about a best practices section to distill what happens if you do stuff what way there and explain it a bit more

Reply:
I'd add a word explicitly after inform.

And add a short remark that Pulumi tries to infer sequence from dependency graph.